oak_functions/loader/src/server.rs

```diff
@@ -353,6 +443,98 @@ fn check_export_function_signature(
     }
 }
 
+pub async fn apply_policy<F, S>(
+    policy: Option<Policy>,
+    function: F,
+) -> anyhow::Result<Response<Body>>
+where
+    F: std::marker::Send + 'static + FnOnce() -> S,
+    S: std::future::Future<Output = anyhow::Result<Response<Body>>> + std::marker::Send,
+{
+    match policy {
+        None => function().await,
+        Some(policy) => {
+            // Use tokio::spawn to actually run the tasks in parallel, for more accurate measurement
+            // of time.
+            let duration = policy.constant_process_time_millis;
+            let timer = tokio::spawn(async move {
+                tokio::time::sleep(Duration::from_millis(duration)).await;
+            });
+            let task = tokio::spawn(async move { function().await });
+
+            // Run the function and the timer tasks in parallel, and measure the time that it
+            // takes for the first task to complete.
+            let start = Instant::now();
+            let response = tokio::select! {
+                _ = timer => Err(anyhow::anyhow!("timeout!")),
+                rsp = task => rsp.unwrap()
+            };
+            let elapsed = start.elapsed().as_millis();
```

Comment:
I was expecting a simpler logic here, something like:

```rust
let task = tokio::spawn(function());
tokio::time::sleep(Duration::from_millis(duration)).await;
match task.now_or_never() {
  Some(v) => ..
  None => ..
}
```

would that work?

Reply:
Yes I think it's fine to postpone anyways, maybe we need >2 threads for this to work? Have you tried with like 4?

Reply:
Yes. This is the error: 
```
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError(())', /workspaces/oak/oak_functions/sdk/test_utils/src/lib.rs:191:19
```

Reply:
That seems to happen because the termination signal is not sent correctly; maybe the sender is dropped before the signal is sent? Note it needs to be explicitly sent in the test case.

Reply:
There are two problems:
1. With multiple threads the client sends the request before the server is up. I could fix this by adding a `sleep`. 
2. I modified https://github.com/project-oak/oak/blob/29a6a2a1300579a292db6e66340491e5f95d2fcc/oak_functions/sdk/test_utils/src/lib.rs#L191-L192 to the following: 
    ```Rust
                    match v {
                        Err(_) => Poll::Pending,
                        Ok(_) => Poll::Ready(()),
                    }
    ```
This fixes the panics, but then again the test fails as before. The policy on processing time is not applied. I have a todo to get back to this when we fix #2026.

PS: I have not pushed this change as it does not really fix the test.